### PR TITLE
Fix malloc_size_of for mac

### PR DESCRIPTION
**Pull request by cheme**
This PR changes the way macos is handled by malloc_size of (in #89 the test was not activated and I did not see the failure for macos).
When no allocator is set macos will use estimate-heapsize (the tests are pretty bad looking but I did not find a better alternative).
Note that I would rather call the system allocator but did not know if the equivalent of 'malloc_size' is exposed anywhere.

If jemalloc is set macos will use jemalloc sizeof (this is the target for parity ethereum).

This PR also allows usage of `MallocSizeOf` trait on `MemoryDB` and mark old `mem_used` method as deprecated.
The related future changes for parity-ethereum would be : https://github.com/paritytech/parity-ethereum/compare/master...cheme:rem_heapsize .

### CODE REVIEW

**Comment by ordian** [comment]
I would personally go for a breaking change approach (I guess `uint`'s version needs to be bumped as well, since a feature is removed? Not sure about `patricia_trie` though).

**Comment by cheme** [comment]
I did not see uint 0.6.0 published on crates.io, so it is probably fine not to increment. For patricia_trie the dependency is a test dependency, so I think there is no need.

File: parity-util-mem/Cargo.toml [hunk 1]
```diff
@@ -17,10 +17,10 @@ elastic-array = { version = "*", optional = true }
 ethereum-types = { version = "*", optional = true }
 parking_lot = { version = "*", optional = true }
 
-[target."cfg(windows)".dependencies.winapi]
+[target.'cfg(target_os = "windows")'.dependencies.winapi]
```
Comment (ordian) [inline parity-util-mem/Cargo.toml]: cfg(windows) should actually work according to [reference](https://doc.rust-lang.org/reference/conditional-compilation.html#set-configuration-options) (equivalent to `target_family = "windows"`), but I prefer `target_os` for consistency with other OSs :+1:

File: parity-util-mem/src/malloc_size.rs [hunk 2]
```diff
@@ -200,27 +200,61 @@ pub trait MallocConditionalShallowSizeOf {
...
+
+impl<T> MallocShallowSizeOf for Vec<T> {
+    fn shallow_size_of(&self, ops: &mut MallocSizeOfOps) -> usize {
+        unsafe { ops.malloc_size_of(self.as_ptr()) }
+    }
+}
+
+// currently this seems only fine with jemalloc
+#[cfg(feature = "std")]
+#[cfg(any(prefixed_jemalloc, target_os = "macos", target_os = "ios", target_os = "android", feature = "jemalloc-global"))]
```
Comment (ordian) [inline parity-util-mem/src/malloc_size.rs]: unrelated, but I think prefixed_jemalloc is coming [from heapsize](https://github.com/servo/heapsize/blob/32615cb931b4871b24f72114cca9faa8bca48399/build.rs#L34-L36) to support old very rust and we shouldn't use it.

File: parity-util-mem/src/allocators.rs [hunk 3]
```diff
@@ -26,6 +26,11 @@
 //!	 - jemalloc: use jemallocator crate
 //!	 - weealloc: default to `estimate_size`
 //!	 - dlmalloc: default to `estimate_size`
+//! - arch x86/macos:
+//!	 - no features: use default alloc, requires using `estimate_size`
```
Comment (ordian) [inline parity-util-mem/src/allocators.rs]: why does it require `estimate_size`, isn't `MallocSizeOf` supposed to work on macOS?

File: parity-util-mem/src/malloc_size.rs (hunk 2 above)
Comment (ordian) [reply parity-util-mem/src/malloc_size.rs]: cc @NikVolf https://github.com/servo/heapsize/pull/44

File: parity-util-mem/src/allocators.rs (hunk 3 above)
Comment (cheme) [reply parity-util-mem/src/allocators.rs]: it is the building failure that I did fail to see in previous commit (test was not activated in ci), it did not find malloc_size_of symbol (or jemalloc if prefixed), I would like to find a way to call the default allocator size of primitive but I do not know any.

Comment (ordian) [reply parity-util-mem/src/allocators.rs]: shouldn't we use `unprefixed_malloc_on_supported_platforms` on [jemallocator](https://github.com/alexcrichton/jemallocator/tree/master/jemalloc-sys#cargo-features)?

**Comment by cheme** [comment]
Since memory-db and patriciatrie has moved to `trie`, this pr no longer allows usage of MallocSizeOf trait on MemoryDB and mark old mem_used method as deprecated.
A PR in `trie` will be needed.

**Comment by ordian** [comment]
Just discovered https://github.com/Aeledfyr/deepsize (similar to estimate-heapsize), although it seems to be unmaintained.

**Pull request merged**